openid-4-verifiable-presentations-1_0.md

```diff
@@ -542,6 +545,266 @@ The Wallet then validates the request as specified in OAuth 2.0 [@RFC6749].
 
 If the Verifier responds with any HTTP error response, the Wallet MUST terminate the process.
 
+# Verifiable Presentation Query Language {#vp_query}
+
+The Verifiable Presentation Query Language (VP Query) is a JSON-encoded query
+language that allows the Verifier to request the Wallet to present Verifiable
+Credentials that match the query. The query consists of a set of constraints
+that the requested Verifiable Credentials must satisfy. The Wallet evaluates the
+query against the Verifiable Credentials it holds and returns the Verifiable
+Presentations that match the query.
+
+A valid VP Query is defined as a JSON-encoded object with the following
+properties:
+
+`credentials`:
+: REQUIRED. An object. Each key in the object represents one
+Credential Query as defined in (#credential_query). The key is a string that
+uniquely identifies the Credential Query within the VP Query.
+
+`rules`:
+: OPTIONAL. An object defining rules that specify which combinations of
+credentials MAY be returned by the Wallet, as defined in (#credential_rules). If
+omitted, all credentials listed in `credentials` are requested for presentation.
+
+## Credential Query {#credential_query}
+
+A Credential Query is an object representing a request for one specific
+credential. It contains the following properties:
+
+`format`:
+: REQUIRED. A string that specifies the format of the requested
+Verifiable Credential. Valid Credential format identifier values are defined in
+Appendix A of [@!OpenID.VCI]. The value of this property MUST be one of the
+supported Credential formats as defined in the Wallet's metadata.
+
+Format-specific properties (using the format value as a key):
+: OPTIONAL.
+Additional properties requested by the Verifier that are specific to the
+Credential format, as defined in (#format_specific_properties). If omitted,
+no format-specific restrictions are placed on the requested Credential.
+
+`purpose`:
+: OPTIONAL. A purpose description array as defined in
+(#purpose_description).
+
+`claims`:
+: OPTIONAL. An array that specifies the claims that the requested
+Verifiable Credential must contain. Each entry in the array corresponds to at
+least one claim that MUST be returned by the Wallet to fulfill the credential
+request (unless specified otherwise by extensions to this specification). Each
+entry MUST be one of the following objects:
+
+- Claim query: An object describing a single claim that is to be returned, as
+  specified in (#claim_query).
+- An optionality rule: An object containing the following properties:
+  - `required` (REQUIRED): defines a number of claims that must be returned from
+    the `from` set of claims, as specified next.
+  - `from` (REQUIRED): An array; each entry is either a claim query or another
+    optionality rule. The entries are listed in order of preference, with the
+    first entry being the most preferred.
+
+More than two nested levels of optionality rules MUST NOT be used by a Verifier
```

Comment:
if credentials have >2 layers of nesting what's the solution?

Reply:
This works for any depth of nesting; the optionality rules are just capped to enable easier processing, but this doesn't reduce expressiveness of the language.

Reply:
Does "two nested levels of optionality rules" mean that there is a total of 2 or 3 instances of "required" in a single "claim" element?